File: OrbitalMechanics/CelestialObjects/Moon.vb
Imports Core
Imports Entities
Imports OrbitalMechanics.Classes

Namespace CelestialObjects
    Public Class Moon
        Inherits OrbitingCelestialObjectBase
        Implements ISphere

        Public Sub New(name As String,
                       mass As Integer,
                       texture As String,
                       radius As Double,
                       entityManager As IEntityManager)

            MyBase.New(name, mass, texture, entityManager)
            _radius = radius
        End Sub

        Private ReadOnly _radius As Double
        Public ReadOnly Property Radius As Double Implements ISphere.Radius
            Get
                Return _radius
            End Get
        End Property

        Public ReadOnly Property AxialTilt As Integer Implements ISphere.AxialTilt
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Public ReadOnly Property RotationSpeed As Integer Implements ISphere.RotationSpeed
            Get
                Throw New NotImplementedException()
            End Get
        End Property

        Private _volume As Double = 0

        Public ReadOnly Property Volume As Double Implements I3DObject.Volume
            Get
                If Double.Equals(_volume, 0.0) AndAlso Radius > 0 Then
                    _volume = Helpers.Shapes.ShapeHelper.VolumeOfASphere(Radius)
                End If
                Return _volume
            End Get
        End Property

    End Class
End Namespace

File: space-game/OrbitalMechanics/CelestialObjects/Moon.cs
using Core.Classes;
using Data.Classes;
using Entities;
using OrbitalMechanics.Interfaces;
using System;
using System.Collections.Generic;

namespace OrbitalMechanics.CelestialObjects
{
    public class Moon : OrbitingCelestialObjectBase, ISphere
    {
        public Moon(string name, Mass mass, List<Texture> textures, Distance radius, IEntityManager entityManager, RingData rings) : base(name, mass, textures, entityManager, rings)
        {
            Radius = radius;
        }


        public Moon(PhysicalData physicalData, IEntityManager entityManager) : base(physicalData.Name, physicalData.Mass, physicalData.Textures, entityManager, physicalData.Rings)
        {
            Radius = physicalData.Radius;
        }

        public Distance Radius { get; }

        public int AxialTilt
        {
            get
            {
                throw new NotImplementedException();
            }
        }

        public int RotationSpeed
        {
            get
            {
                throw new NotImplementedException();
            }
        }

        private double _volume = 0;
        public double Volume
        {
            get
            {
                if (double.Equals(_volume, 0.0) && Radius.Kilometers > 0)
                {
                    _volume = Core.Helpers.Shapes.ShapeHelper.VolumeOfASphere(Radius.Kilometers);
                }
                return _volume;
            }
        }
    }
}
